File: project-05-db-migration/code-java/src/test/java/com/example/demo/service/MigrationServiceTest.java
package com.example.demo.service;

import com.example.jooq.legacy.tables.records.OrdersV1Record;
import com.example.jooq.legacy.tables.records.UsersV1Record;
import com.example.jooq.target.tables.records.MemberRecord;
import com.example.jooq.target.tables.records.PurchaseOrderRecord;
import org.jooq.DSLContext;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.test.context.SpringBootTest;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import static com.example.jooq.legacy.Tables.ORDERS_V1;
import static com.example.jooq.legacy.Tables.USERS_V1;
import static com.example.jooq.target.Tables.MEMBER;
import static com.example.jooq.target.Tables.PURCHASE_ORDER;
import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest
class MigrationServiceTest {

    @Autowired
    private MigrationService migrationService;

    @Autowired
    @Qualifier("legacyDsl")
    private DSLContext legacyDsl;

    @Autowired
    @Qualifier("targetDsl")
    private DSLContext targetDsl;

    @BeforeEach
    void setUp() {
        // Clear both databases
        legacyDsl.truncate(ORDERS_V1).cascade().execute();
        legacyDsl.truncate(USERS_V1).cascade().execute();
        
        targetDsl.truncate(PURCHASE_ORDER).cascade().execute();
        targetDsl.truncate(MEMBER).cascade().execute();

        // 1. Generate 100 Legacy Users
        List<UsersV1Record> legacyUsers = new ArrayList<>();
        for (int i = 1; i <= 100; i++) {
            UsersV1Record user = legacyDsl.newRecord(USERS_V1);
            user.setId((long) i);
            user.setFullName("User FirstName" + i + " LastName" + i);
            user.setUserStatus(i % 2 == 0 ? "VALID" : "REVOKED"); // Even: VALID (ACTIVE), Odd: REVOKED (INACTIVE)
            user.setCreatedDate(LocalDateTime.now().minusDays(i));
            legacyUsers.add(user);
        }
        legacyDsl.batchInsert(legacyUsers).execute();

        // 2. Generate 100 Legacy Orders
        List<OrdersV1Record> legacyOrders = new ArrayList<>();
        for (int i = 1; i <= 100; i++) {
            OrdersV1Record order = legacyDsl.newRecord(ORDERS_V1);
            order.setId((long) i);
            order.setUserId((long) i);
            order.setOrderTotal(new BigDecimal("10.50").multiply(new BigDecimal(i)));
            order.setState(i % 3 == 0 ? "PAID" : "FAIL"); // PAID -> COMPLETED, FAIL -> PENDING
            legacyOrders.add(order);
        }
        legacyDsl.batchInsert(legacyOrders).execute();
    }

    @Test
    @DisplayName("Should successfully migrate all users and orders with correct type mappings")
    void testMigrateAll() {
        // When
        migrationService.migrateAll();

        // Then: verify Members
        List<MemberRecord> members = targetDsl.selectFrom(MEMBER).fetch();
        assertThat(members).hasSize(100);

        MemberRecord member2 = members.stream().filter(m -> m.getId() == 2L).findFirst().orElseThrow();
        assertThat(member2.getFirstName()).isEqualTo("User");
        assertThat(member2.getLastName()).isEqualTo("FirstName2 LastName2");
        assertThat(member2.getStatus()).isEqualTo("ACTIVE"); // mapped from VALID

        MemberRecord member3 = members.stream().filter(m -> m.getId() == 3L).findFirst().orElseThrow();
        assertThat(member3.getStatus()).isEqualTo("INACTIVE"); // mapped from REVOKED

        // Then: verify Purchase Orders
        List<PurchaseOrderRecord> pos = targetDsl.selectFrom(PURCHASE_ORDER).fetch();
        assertThat(pos).hasSize(100);

        PurchaseOrderRecord po3 = pos.stream().filter(p -> p.getId() == 3L).findFirst().orElseThrow();
        assertThat(po3.getMemberId()).isEqualTo(3L);
        assertThat(po3.getOrderState()).isEqualTo("COMPLETED"); // mapped from PAID

        PurchaseOrderRecord po4 = pos.stream().filter(p -> p.getId() == 4L).findFirst().orElseThrow();
        assertThat(po4.getOrderState()).isEqualTo("PENDING"); // mapped from FAIL
    }
}

File: project-05-db-migration/code-kotlin/src/test/kotlin/com/example/demo/service/MigrationServiceTest.kt
package com.example.demo.service

import com.example.jooq.legacy.tables.references.ORDERS_V1
import com.example.jooq.legacy.tables.references.USERS_V1
import com.example.jooq.target.tables.references.MEMBER
import com.example.jooq.target.tables.references.PURCHASE_ORDER
import org.assertj.core.api.Assertions.assertThat
import org.jooq.DSLContext
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.boot.test.context.SpringBootTest
import java.math.BigDecimal
import java.time.LocalDateTime

@SpringBootTest
class MigrationServiceTest(
    @Autowired private val migrationService: MigrationService,
    @Autowired @Qualifier("legacyDsl") private val legacyDsl: DSLContext,
    @Autowired @Qualifier("targetDsl") private val targetDsl: DSLContext
) {

    @BeforeEach
    fun setUp() {
        // Clear
        legacyDsl.truncate(ORDERS_V1).cascade().execute()
        legacyDsl.truncate(USERS_V1).cascade().execute()
        
        targetDsl.truncate(PURCHASE_ORDER).cascade().execute()
        targetDsl.truncate(MEMBER).cascade().execute()

        // Generate 100 Legacy Users
        val legacyUsers = (1..100).map { i ->
            legacyDsl.newRecord(USERS_V1).apply {
                id = i.toLong()
                fullName = "KotlinUser FirstName$i LastName$i"
                userStatus = if (i % 2 == 0) "VALID" else "REVOKED"
                createdDate = LocalDateTime.now().minusDays(i.toLong())
            }
        }
        legacyDsl.batchInsert(legacyUsers).execute()

        // Generate 100 Legacy Orders
        val legacyOrders = (1..100).map { i ->
            legacyDsl.newRecord(ORDERS_V1).apply {
                id = i.toLong()
                userId = i.toLong()
                orderTotal = BigDecimal("10.50").multiply(BigDecimal(i))
                state = if (i % 3 == 0) "PAID" else "FAIL"
            }
        }
        legacyDsl.batchInsert(legacyOrders).execute()
    }

    @Test
    @DisplayName("Should successfully migrate all users and orders with correct type mappings")
    fun testMigrateAll() {
        // When
        migrationService.migrateAll()

        // Then: Members
        val members = targetDsl.selectFrom(MEMBER).fetch()
        assertThat(members).hasSize(100)

        val member2 = members.first { it.id == 2L }
        assertThat(member2.firstName).isEqualTo("KotlinUser")
        assertThat(member2.lastName).isEqualTo("FirstName2 LastName2")
        assertThat(member2.status).isEqualTo("ACTIVE")

        val member3 = members.first { it.id == 3L }
        assertThat(member3.status).isEqualTo("INACTIVE")

        // Then: Purchase Orders
        val pos = targetDsl.selectFrom(PURCHASE_ORDER).fetch()
        assertThat(pos).hasSize(100)

        val po3 = pos.first { it.id == 3L }
        assertThat(po3.memberId).isEqualTo(3L)
        assertThat(po3.orderState).isEqualTo("COMPLETED")

        val po4 = pos.first { it.id == 4L }
        assertThat(po4.orderState).isEqualTo("PENDING")
    }
}
